src/Storages/System/StorageSystemTables.cpp

```diff
@@ -455,6 +458,33 @@ class TablesBlockSource : public ISource
                         res_columns[res_index++]->insertDefault();
                 }
 
+                if (columns_mask[src_index++]) 
+                {
+                    auto parts = table->totalParts(settings);
+                    if (parts)
+                        res_columns[res_index++]->insert(*parts);
+                    else
+                        res_columns[res_index++]->insertDefault();
+                }
+
+                if (columns_mask[src_index++]) 
+                {
+                    auto active_parts = table->totalActiveParts(settings);
+                    if (active_parts)
+                        res_columns[res_index++]->insert(*active_parts);
+                    else
+                        res_columns[res_index++]->insertDefault();
+                }
+
+                if (columns_mask[src_index++]) 
+                {
+                    auto total_marks = table->totalMarks(settings);
+                    if (total_marks)
+                        res_columns[res_index++]->insert(*total_marks);
+                    else
+                        res_columns[res_index++]->insertDefault();
+                }
+
```

Comment:
We can `dynamic_cast` the `table` to `MergeTreeData` and compute it directly with method from `MergeTreeData`.